src/OrleansSQLUtils/CreateOrleansTables_SqlServer.sql

```diff
@@ -641,6 +727,161 @@ VALUES
+			ServiceId,
+			PayloadBinary,
+			PayloadJson,
+			PayloadXml,
+			ModifiedOn,
+			Version
+		)
+		SELECT
+			@GrainIdHash,
+			@GrainIdString,
+			@GrainTypeHash,
+			@GrainTypeString,
+			@ServiceId,
+			@PayloadBinary,
+			@PayloadJson,
+			@PayloadXml,
+			GETUTCDATE(),
+			1
+		 WHERE NOT EXISTS
+		 (
+			-- There should not be any version of this grain state.
+			SELECT 1
+			FROM Storage WITH(UPDLOCK, HOLDLOCK)
+			WHERE
+				GrainIdHash = @GrainIdHash AND @GrainIdHash IS NOT NULL
+				AND GrainTypeHash = @GrainTypeHash AND @GrainTypeHash IS NOT NULL
+				AND GrainIdString = @GrainIdString AND @GrainIdString IS NOT NULL
+				AND GrainTypeString = @GrainTypeString AND @GrainTypeString IS NOT NULL
+				AND ServiceId = @ServiceId AND @ServiceId IS NOT NULL
+		 ) OPTION(FAST 1, OPTIMIZE FOR(@GrainIdHash UNKNOWN, @GrainTypeHash UNKNOWN));
+
+		IF @@ROWCOUNT > 0
+		BEGIN
+			SET @NewGrainStateVersion = 1;
+		END
+	END
+
+	SELECT @NewGrainStateVersion AS NewGrainStateVersion;
+	COMMIT TRANSACTION;'
+);
+
+
+INSERT INTO OrleansQuery(QueryKey, QueryText)
+VALUES
+(
+	'ClearStorageKey',
+	'BEGIN TRANSACTION;
+    SET XACT_ABORT, NOCOUNT ON;
+    DECLARE @NewGrainStateVersion AS INT = @GrainStateVersion;
+    UPDATE Storage
+    SET
+	    PayloadBinary = NULL,
+	    PayloadJson = NULL,
+	    PayloadXml = NULL,
+	    ModifiedOn = GETUTCDATE(),
+	    Version = Version + 1,
+        @NewGrainStateVersion = Version + 1
+    WHERE
+	    GrainIdHash = @GrainIdHash AND @GrainIdHash IS NOT NULL
+        AND GrainTypeHash = @GrainTypeHash AND @GrainTypeHash IS NOT NULL
+        AND GrainIdString = @GrainIdString AND @GrainIdString IS NOT NULL
+        AND GrainTypeString = @GrainTypeString AND @GrainTypeString IS NOT NULL
+		AND ServiceId = @ServiceId AND @ServiceId IS NOT NULL
+        AND Version IS NOT NULL AND Version = @GrainStateVersion AND @GrainStateVersion IS NOT NULL
+        OPTION(FAST 1, OPTIMIZE FOR(@GrainIdHash UNKNOWN, @GrainTypeHash UNKNOWN));
+
+    SELECT @NewGrainStateVersion;
+    COMMIT TRANSACTION;'
+);
+
+
+INSERT INTO OrleansQuery(QueryKey, QueryText)
+VALUES
+(
+	'ReadFromStorageKey',
+	'-- The application code will deserialize the relevant result. Not that the query optimizer
+    -- estimates the result of rows based on its knowledge on the index. It does not know there
+    -- will be only one row returned. Forcing the optimizer to process the first found row quickly
+    -- creates an estimate for a one-row result and makes a difference on multi-million row tables.
+    -- Also the optimizer is instructed to always use the same plan via index using the OPTIMIZE
+    -- FOR UNKNOWN flags. These hints are only available in SQL Server 2008 and later. They
+    -- should guarantee the execution time is robustly basically the same from query-to-query.
+    SELECT
+        PayloadBinary,
+        PayloadXml,
+        PayloadJson,
+        Version
+    FROM
+        Storage
+    WHERE
+        GrainIdHash = @GrainIdHash
```

Comment:
I agree on that, but I have a bit of a hard time on setting up a `deploymentId` using the `StorageProviderManager` interface. The only way I know how to do that would be to make it mutable in `RelationalStorageProvider`.

One should somehow sneak it at
```
public CommonFixture()
{
    DefaultProviderRuntime = new StorageProviderManager(new GrainFactory(), null);
    ((StorageProviderManager)DefaultProviderRuntime).LoadEmptyStorageProviders(new ClientProviderRuntime(new GrainFactory(), null)).WaitWithThrow(TestConstants.InitTimeout);

    SerializationManager.InitializeForTesting();
}
```
instead of that `null` in `StorageProviderManager` constructor should be something that can provide it at one's leisure.